Amazon.Ion.ObjectMapper/IonObjectSerializer.cs

```diff
@@ -22,47 +22,75 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public object Deserialize(IIonReader reader)
         {
-            var targetObject = options.ObjectFactory.Create(options, reader, targetType);
-            reader.StepIn();
-
-            IonType ionType;
-            while ((ionType = reader.MoveNext()) != IonType.None)
+            if (reader == null)
+            {
+                throw new InvalidOperationException("Cannot deserialize with a null Ion reader");
+            }
+            
+            try
             {
-                var property = FindProperty(reader.CurrentFieldName);
-                FieldInfo field;
-                if (property != null)
+                reader.StepIn();
+                
+                var ionConstructors = targetType.GetConstructors().Where(IsIonConstructor);
+                switch (ionConstructors.Count())
```

Comment:
`case` seems a bit complex for what's needed, how about using `1 != ionConstructors.Count()` ?

Reply:
`1 != ionConstructors.Count()` isn't enough as we still need to distinguish between count = 0 and count > 1.

Take a look at what I tried instead.